pipeline/src/node_properties/camera_properties.rs

```diff
@@ -0,0 +1,113 @@
+use crate::resolution::Resolution;
+
+pub trait CameraProperties {
+    type Runtime: CameraRuntime;
+
+    fn uri(&self) -> &url::Url;
+    fn set_uri(&mut self, url: url::Url);
+    fn resolution(&self) -> Option<&Resolution>;
+    fn set_resolution(&mut self, resolution: Option<Resolution>);
+    fn framerate(&self) -> Option<u32>;
+    fn set_framerate(&mut self, framerate: Option<u32>);
+    fn runtime(&self) -> &Self::Runtime;
+    fn runtime_mut(&mut self) -> &mut Self::Runtime;
+}
+
+pub trait CameraRuntime {
+    fn source(&self) -> Option<&String>;
+    fn set_source(&mut self, source: Option<String>);
+}
+
+macro_rules! impl_camera_props {
+    ($camera_type:ty, $runtime: ty) => {
+        impl crate::CameraProperties for $camera_type {
+            type Runtime = $runtime;
+
+            fn uri(&self) -> &url::Url {
+                &self.uri
+            }
+
+            fn set_uri(&mut self, uri: url::Url) {
+                self.uri = uri
+            }
+
+            fn resolution(&self) -> Option<&Resolution> {
+                self.resolution.as_ref()
+            }
+
+            fn set_resolution(&mut self, resolution: Option<Resolution>) {
+                self.resolution = resolution;
+            }
+
+            fn framerate(&self) -> Option<u32> {
+                self.framerate
+            }
+
+            fn set_framerate(&mut self, framerate: Option<u32>) {
+                self.framerate = framerate;
+            }
+
+            fn runtime(&self) -> &$runtime {
+                &self.runtime
+            }
+
+            fn runtime_mut(&mut self) -> &mut $runtime {
+                &mut self.runtime
+            }
+        }
+
+        impl crate::CameraRuntime for $runtime {
+            // FIXME: This should ideally return &Option<&str>
```

Comment:
I'm guessing this means `Option<&str>`? Seems to just require changing the definition from `as_ref()` to `as_deref()`.

Reply:
Thanks, I'll try that.